File: Enemies.lua
--[[--------------------------------------------------------------------
    Copyright (C) 2012 Sidoine De Wispelaere.
    Copyright (C) 2012, 2013, 2014 Johnny C. Lam.
    See the file LICENSE.txt for copying permission.
--]]--------------------------------------------------------------------

-- Gather information about ennemies

local OVALE, Ovale = ...
local OvaleEnemies = Ovale:NewModule("OvaleEnemies", "AceEvent-3.0", "AceTimer-3.0")
Ovale.OvaleEnemies = OvaleEnemies

--<private-static-properties>
local L = Ovale.L
local OvaleDebug = Ovale.OvaleDebug
local OvaleProfiler = Ovale.OvaleProfiler

-- Forward declarations for module dependencies.
local OvaleGUID = nil
local OvaleState = nil

local bit_band = bit.band
local bit_bor = bit.bor
local ipairs = ipairs
local pairs = pairs
local strfind = string.find
local tostring = tostring
local wipe = wipe
local API_GetTime = GetTime
local COMBATLOG_OBJECT_AFFILIATION_MINE = COMBATLOG_OBJECT_AFFILIATION_MINE
local COMBATLOG_OBJECT_AFFILIATION_PARTY = COMBATLOG_OBJECT_AFFILIATION_PARTY
local COMBATLOG_OBJECT_AFFILIATION_RAID = COMBATLOG_OBJECT_AFFILIATION_RAID
local COMBATLOG_OBJECT_REACTION_FRIENDLY = COMBATLOG_OBJECT_REACTION_FRIENDLY

-- Bitmask for player, party or raid unit controller affiliation.
local GROUP_MEMBER = bit_bor(COMBATLOG_OBJECT_AFFILIATION_MINE, COMBATLOG_OBJECT_AFFILIATION_PARTY, COMBATLOG_OBJECT_AFFILIATION_RAID)

-- Register for debugging messages.
OvaleDebug:RegisterDebugging(OvaleEnemies)
-- Register for profiling.
OvaleProfiler:RegisterProfiling(OvaleEnemies)

--[[
	List of CLEU event suffixes that can correspond to the player damaging or try to
	damage (tag) an enemy, or vice versa.
--]]
local CLEU_TAG_SUFFIXES = {
	"_DAMAGE",
	"_MISSED",
	"_AURA_APPLIED",
	"_AURA_APPLIED_DOSE",
	"_AURA_REFRESH",
	"_CAST_START",
	"_INTERRUPT",
	"_DISPEL",
	"_DISPEL_FAILED",
	"_STOLEN",
	"_DRAIN",
	"_LEECH",
}

-- Table of CLEU events for auto-attacks.
local CLEU_AUTOATTACK = {
	RANGED_DAMAGE = true,
	RANGED_MISSED = true,
	SWING_DAMAGE = true,
	SWING_MISSED = true,
}

-- Table of CLEU events for when a unit is removed from combat.
local CLEU_UNIT_REMOVED = {
	UNIT_DESTROYED = true,
	UNIT_DIED = true,
	UNIT_DISSIPATES = true,
}

-- Player's GUID.
local self_playerGUID = nil

-- enemyName[guid] = name
local self_enemyName = {}
-- enemyLastSeen[guid] = timestamp
local self_enemyLastSeen = {}
-- taggedEnemyLastSeen[guid] = timestamp
-- GUIDs used as keys for this table are a subset of the GUIDs used for enemyLastSeen.
local self_taggedEnemyLastSeen = {}

-- Timer for reaper function to remove inactive enemies.
local self_reaperTimer = nil
local REAP_INTERVAL = 3
--</private-static-properties>

--<public-static-properties>
-- Total number of active enemies.
OvaleEnemies.activeEnemies = 0
-- Total number of tagged enemies.
OvaleEnemies.taggedEnemies = 0
--</public-static-properties>

--<private-static-methods>
local function IsTagEvent(cleuEvent)
	local isTagEvent = false
	if CLEU_AUTOATTACK[cleuEvent] then
		isTagEvent = true
	else
		for _, suffix in ipairs(CLEU_TAG_SUFFIXES) do
			if strfind(cleuEvent, suffix .. "$") then
				isTagEvent = true
				break
			end
		end
	end
	return isTagEvent
end

local function IsFriendly(unitFlags, isGroupMember)
	return bit_band(unitFlags, COMBATLOG_OBJECT_REACTION_FRIENDLY) > 0 and (not isGroupMember or bit_band(unitFlags, GROUP_MEMBER) > 0)
end
--</private-static-methods>

--<public-static-methods>
function OvaleEnemies:OnInitialize()
	-- Resolve module dependencies.
	OvaleGUID = Ovale.OvaleGUID
	OvaleState = Ovale.OvaleState
end

function OvaleEnemies:OnEnable()
	self_playerGUID = Ovale.playerGUID
	if not self_reaperTimer then
		self_reaperTimer = self:ScheduleRepeatingTimer("RemoveInactiveEnemies", REAP_INTERVAL)
	end
	self:RegisterEvent("COMBAT_LOG_EVENT_UNFILTERED")
	self:RegisterEvent("PLAYER_REGEN_DISABLED")
	OvaleState:RegisterState(self, self.statePrototype)
end

function OvaleEnemies:OnDisable()
	OvaleState:UnregisterState(self)
	if not self_reaperTimer then
		self:CancelTimer(self_reaperTimer)
		self_reaperTimer = nil
	end
	self:UnregisterEvent("COMBAT_LOG_EVENT_UNFILTERED")
	self:UnregisterEvent("PLAYER_REGEN_DISABLED")
end

function OvaleEnemies:COMBAT_LOG_EVENT_UNFILTERED(event, timestamp, cleuEvent, hideCaster, sourceGUID, sourceName, sourceFlags, sourceRaidFlags, destGUID, destName, destFlags, destRaidFlags, ...)
	if CLEU_UNIT_REMOVED[cleuEvent] then
		local now = API_GetTime()
		self:RemoveEnemy(cleuEvent, destGUID, now, true)
	elseif sourceGUID and sourceGUID ~= "" and sourceName and sourceFlags and destGUID and destGUID ~= "" and destName and destFlags then
		if not IsFriendly(sourceFlags) and IsFriendly(destFlags, true) then
			--[[
				Unfriendly enemy attacks friendly group member.
				Filter out periodic damage effects since they can occur after the caster's death
				and can cause false positives.
			--]]
			if not cleuEvent == "SPELL_PERIODIC_DAMAGE" and IsTagEvent(cleuEvent) then
				local now = API_GetTime()
				self:AddEnemy(cleuEvent, sourceGUID, sourceName, now)
			end
		elseif IsFriendly(sourceFlags, true) and not IsFriendly(destFlags) and IsTagEvent(cleuEvent) then
			-- Friendly group member attacks unfriendly enemy.
			local now = API_GetTime()
			-- Treat both player and pet attacks as a player tag.
			local isPlayerTag = (sourceGUID == self_playerGUID) or OvaleGUID:IsPlayerPet(sourceGUID)
			self:AddEnemy(cleuEvent, destGUID, destName, now, isPlayerTag)
		end
	end
end

function OvaleEnemies:PLAYER_REGEN_DISABLED()
	-- Reset enemy tracking when combat starts.
	wipe(self_enemyName)
	wipe(self_enemyLastSeen)
	wipe(self_taggedEnemyLastSeen)
	self.activeEnemies = 0
	self.taggedEnemies = 0
end

--[[
	Remove enemies that have been inactive for at least REAP_INTERVAL seconds.
	These enemies are not in combat with your group, out of range, or
	incapacitated and shouldn't count toward the number of active enemies.
--]]
function OvaleEnemies:RemoveInactiveEnemies()
	self:StartProfiling("OvaleEnemies_RemoveInactiveEnemies")
	local now = API_GetTime()
	-- Remove inactive enemies first.
	for guid, timestamp in pairs(self_enemyLastSeen) do
		if now - timestamp > REAP_INTERVAL then
			self:RemoveEnemy("REAPED", guid, now)
		end
	end
	-- Remove inactive tagged enemies last.
	for guid, timestamp in pairs(self_taggedEnemyLastSeen) do
		if now - timestamp > REAP_INTERVAL then
			self:RemoveTaggedEnemy("REAPED", guid, now)
		end
	end
	self:StopProfiling("OvaleEnemies_RemoveInactiveEnemies")
end

function OvaleEnemies:AddEnemy(cleuEvent, guid, name, timestamp, isTagged)
	self:StartProfiling("OvaleEnemies_AddEnemy")
	if guid then
		self_enemyName[guid] = name
		local changed = false
		do
			-- Update last time this enemy was seen.
			if not self_enemyLastSeen[guid] then
				self.activeEnemies = self.activeEnemies + 1
				changed = true
			end
			self_enemyLastSeen[guid] = timestamp
		end
		if isTagged then
			-- Update last time this enemy was tagged.
			if not self_taggedEnemyLastSeen[guid] then
				self.taggedEnemies = self.taggedEnemies + 1
				changed = true
			end
			self_taggedEnemyLastSeen[guid] = timestamp
		end
		if changed then
			self:DebugTimestamp("%s: %d/%d enemy seen: %s (%s)", cleuEvent, self.taggedEnemies, self.activeEnemies, guid, name)
			Ovale.refreshNeeded[self_playerGUID] = true
		end
	end
	self:StopProfiling("OvaleEnemies_AddEnemy")
end

function OvaleEnemies:RemoveEnemy(cleuEvent, guid, timestamp, isDead)
	self:StartProfiling("OvaleEnemies_RemoveEnemy")
	if guid then
		local name = self_enemyName[guid]
		local changed = false
		-- Update seen enemy count.
		if self_enemyLastSeen[guid] then
			self_enemyLastSeen[guid] = nil
			if self.activeEnemies > 0 then
				self.activeEnemies = self.activeEnemies - 1
				changed = true
			end
		end
		-- Update tagged enemy count.
		if self_taggedEnemyLastSeen[guid] then
			self_taggedEnemyLastSeen[guid] = nil
			if self.taggedEnemies > 0 then
				self.taggedEnemies = self.taggedEnemies - 1
				changed = true
			end
		end
		if changed then
			self:DebugTimestamp("%s: %d/%d enemy %s: %s (%s)", cleuEvent, self.taggedEnemies, self.activeEnemies, isDead and "died" or "removed", guid, name)
			Ovale.refreshNeeded[self_playerGUID] = true
			self:SendMessage("Ovale_InactiveUnit", guid, isDead)
		end
	end
	self:StopProfiling("OvaleEnemies_RemoveEnemy")
end

function OvaleEnemies:RemoveTaggedEnemy(cleuEvent, guid, timestamp)
	self:StartProfiling("OvaleEnemies_RemoveTaggedEnemy")
	if guid then
		local name = self_enemyName[guid]
		local tagged = self_taggedEnemyLastSeen[guid]
		if tagged then
			self_taggedEnemyLastSeen[guid] = nil
			if self.taggedEnemies > 0 then
				self.taggedEnemies = self.taggedEnemies - 1
			end
			self:DebugTimestamp("%s: %d/%d enemy removed: %s (%s), last tagged at %f", cleuEvent, self.taggedEnemies, self.activeEnemies, guid, name, tagged)
			Ovale.refreshNeeded[self_playerGUID] = true
		end
	end
	self:StopProfiling("OvaleEnemies_RemoveEnemy")
end

function OvaleEnemies:DebugEnemies()
	for guid, seen in pairs(self_enemyLastSeen) do
		local name = self_enemyName[guid]
		local tagged = self_taggedEnemyLastSeen[guid]
		if tagged then
			self:Print("Tagged enemy %s (%s) last seen at %f", guid, name, tagged)
		else
			self:Print("Enemy %s (%s) last seen at %f", guid, name, seen)
		end
	end
	self:Print("Total enemies: %d", self.activeEnemies)
	self:Print("Total tagged enemies: %d", self.taggedEnemies)
end
--</public-static-methods>

--[[----------------------------------------------------------------------------
	State machine for simulator.
--]]----------------------------------------------------------------------------

--<public-static-properties>
OvaleEnemies.statePrototype = {}
--</public-static-properties>

--<private-static-properties>
local statePrototype = OvaleEnemies.statePrototype
--</private-static-properties>

--<state-properties>
-- Total number of active enemies.
statePrototype.activeEnemies = nil
-- Total number of tagged enemies.
statePrototype.taggedEnemies = nil
-- Requested number of enemies.
statePrototype.enemies = nil
--</state-properties>

--<public-static-methods>
-- Initialize the state.
function OvaleEnemies:InitializeState(state)
	state.enemies = nil
end

-- Reset the state to the current conditions.
function OvaleEnemies:ResetState(state)
	self:StartProfiling("OvaleEnemies_ResetState")
	state.activeEnemies = self.activeEnemies
	state.taggedEnemies = self.taggedEnemies
	self:StopProfiling("OvaleEnemies_ResetState")
end

-- Release state resources prior to removing from the simulator.
function OvaleEnemies:CleanState(state)
	state.activeEnemies = nil
	state.taggedEnemies = nil
	state.enemies = nil
end
--</public-static-methods>

File: src/Enemies.ts
import { OvaleDebug } from "./Debug";
import { OvaleProfiler } from "./Profiler";
import { Ovale } from "./Ovale";
import { OvaleGUID } from "./GUID";
import { OvaleState } from "./State";
import aceEvent from "@wowts/ace_event-3.0";
import AceTimer from "@wowts/ace_timer-3.0";
import { band, bor } from "@wowts/bit";
import { ipairs, pairs, wipe, truthy } from "@wowts/lua";
import { find } from "@wowts/string";
import { GetTime, COMBATLOG_OBJECT_AFFILIATION_MINE, COMBATLOG_OBJECT_AFFILIATION_PARTY, COMBATLOG_OBJECT_AFFILIATION_RAID, COMBATLOG_OBJECT_REACTION_FRIENDLY } from "@wowts/wow-mock";

export let OvaleEnemies: OvaleEnemiesClass;
let GROUP_MEMBER = bor(COMBATLOG_OBJECT_AFFILIATION_MINE, COMBATLOG_OBJECT_AFFILIATION_PARTY, COMBATLOG_OBJECT_AFFILIATION_RAID);
let CLEU_TAG_SUFFIXES = {
    1: "_DAMAGE",
    2: "_MISSED",
    3: "_AURA_APPLIED",
    4: "_AURA_APPLIED_DOSE",
    5: "_AURA_REFRESH",
    6: "_CAST_START",
    7: "_INTERRUPT",
    8: "_DISPEL",
    9: "_DISPEL_FAILED",
    10: "_STOLEN",
    11: "_DRAIN",
    12: "_LEECH"
}
let CLEU_AUTOATTACK = {
    RANGED_DAMAGE: true,
    RANGED_MISSED: true,
    SWING_DAMAGE: true,
    SWING_MISSED: true
}
let CLEU_UNIT_REMOVED = {
    UNIT_DESTROYED: true,
    UNIT_DIED: true,
    UNIT_DISSIPATES: true
}
let self_enemyName = {
}
let self_enemyLastSeen = {
}
let self_taggedEnemyLastSeen = {
}
let self_reaperTimer = undefined;
let REAP_INTERVAL = 3;
const IsTagEvent = function(cleuEvent) {
    let isTagEvent = false;
    if (CLEU_AUTOATTACK[cleuEvent]) {
        isTagEvent = true;
    } else {
        for (const [, suffix] of ipairs(CLEU_TAG_SUFFIXES)) {
            if (truthy(find(cleuEvent, `${suffix}$`))) {
                isTagEvent = true;
                break;
            }
        }
    }
    return isTagEvent;
}
const IsFriendly = function(unitFlags, isGroupMember?) {
    return band(unitFlags, COMBATLOG_OBJECT_REACTION_FRIENDLY) > 0 && (!isGroupMember || band(unitFlags, GROUP_MEMBER) > 0);
}

class EnemiesData {
    activeEnemies = 0;
    taggedEnemies = 0;
    enemies = undefined;
}
let OvaleEnemiesBase = OvaleState.RegisterHasState(OvaleDebug.RegisterDebugging(OvaleProfiler.RegisterProfiling(Ovale.NewModule("OvaleEnemies", aceEvent, AceTimer))), EnemiesData);

class OvaleEnemiesClass extends OvaleEnemiesBase {
    
    OnInitialize() {
        if (!self_reaperTimer) {
            self_reaperTimer = this.ScheduleRepeatingTimer("RemoveInactiveEnemies", REAP_INTERVAL);
        }
        this.RegisterEvent("COMBAT_LOG_EVENT_UNFILTERED");
        this.RegisterEvent("PLAYER_REGEN_DISABLED");
    }
    OnDisable() {
        if (!self_reaperTimer) {
            this.CancelTimer(self_reaperTimer);
            self_reaperTimer = undefined;
        }
        this.UnregisterEvent("COMBAT_LOG_EVENT_UNFILTERED");
        this.UnregisterEvent("PLAYER_REGEN_DISABLED");
    }
    COMBAT_LOG_EVENT_UNFILTERED(event, timestamp, cleuEvent, hideCaster, sourceGUID, sourceName, sourceFlags, sourceRaidFlags, destGUID, destName, destFlags, destRaidFlags, ...__args) {
        if (CLEU_UNIT_REMOVED[cleuEvent]) {
            let now = GetTime();
            this.RemoveEnemy(cleuEvent, destGUID, now, true);
        } else if (sourceGUID && sourceGUID != "" && sourceName && sourceFlags && destGUID && destGUID != "" && destName && destFlags) {
            if (!IsFriendly(sourceFlags) && IsFriendly(destFlags, true)) {
                if (!(cleuEvent == "SPELL_PERIODIC_DAMAGE" && IsTagEvent(cleuEvent))) {
                    let now = GetTime();
                    this.AddEnemy(cleuEvent, sourceGUID, sourceName, now);
                }
            } else if (IsFriendly(sourceFlags, true) && !IsFriendly(destFlags) && IsTagEvent(cleuEvent)) {
                let now = GetTime();
                let isPlayerTag = (sourceGUID == Ovale.playerGUID) || OvaleGUID.IsPlayerPet(sourceGUID);
                this.AddEnemy(cleuEvent, destGUID, destName, now, isPlayerTag);
            }
        }
    }
    PLAYER_REGEN_DISABLED() {
        wipe(self_enemyName);
        wipe(self_enemyLastSeen);
        wipe(self_taggedEnemyLastSeen);
        this.current.activeEnemies = 0;
        this.current.taggedEnemies = 0;
    }
    RemoveInactiveEnemies() {
        this.StartProfiling("OvaleEnemies_RemoveInactiveEnemies");
        let now = GetTime();
        for (const [guid, timestamp] of pairs(self_enemyLastSeen)) {
            if (now - timestamp > REAP_INTERVAL) {
                this.RemoveEnemy("REAPED", guid, now);
            }
        }
        for (const [guid, timestamp] of pairs(self_taggedEnemyLastSeen)) {
            if (now - timestamp > REAP_INTERVAL) {
                this.RemoveTaggedEnemy("REAPED", guid, now);
            }
        }
        this.StopProfiling("OvaleEnemies_RemoveInactiveEnemies");
    }
    private AddEnemy(cleuEvent, guid, name, timestamp, isTagged?) {
        this.StartProfiling("OvaleEnemies_AddEnemy");
        if (guid) {
            self_enemyName[guid] = name;
            let changed = false;
            {
                if (!self_enemyLastSeen[guid]) {
                    this.current.activeEnemies = this.current.activeEnemies + 1;
                    changed = true;
                }
                self_enemyLastSeen[guid] = timestamp;
            }
            if (isTagged) {
                if (!self_taggedEnemyLastSeen[guid]) {
                    this.current.taggedEnemies = this.current.taggedEnemies + 1;
                    changed = true;
                }
                self_taggedEnemyLastSeen[guid] = timestamp;
            }
            if (changed) {
                this.DebugTimestamp("%s: %d/%d enemy seen: %s (%s)", cleuEvent, this.current.taggedEnemies, this.current.activeEnemies, guid, name);
                Ovale.needRefresh();
            }
        }
        this.StopProfiling("OvaleEnemies_AddEnemy");
    }
    private RemoveEnemy(cleuEvent, guid, timestamp, isDead?) {
        this.StartProfiling("OvaleEnemies_RemoveEnemy");
        if (guid) {
            let name = self_enemyName[guid];
            let changed = false;
            if (self_enemyLastSeen[guid]) {
                self_enemyLastSeen[guid] = undefined;
                if (this.current.activeEnemies > 0) {
                    this.current.activeEnemies = this.current.activeEnemies - 1;
                    changed = true;
                }
            }
            if (self_taggedEnemyLastSeen[guid]) {
                self_taggedEnemyLastSeen[guid] = undefined;
                if (this.current.taggedEnemies > 0) {
                    this.current.taggedEnemies = this.current.taggedEnemies - 1;
                    changed = true;
                }
            }
            if (changed) {
                this.DebugTimestamp("%s: %d/%d enemy %s: %s (%s)", cleuEvent, this.current.taggedEnemies, this.current.activeEnemies, isDead && "died" || "removed", guid, name);
                Ovale.needRefresh();
                this.SendMessage("Ovale_InactiveUnit", guid, isDead);
            }
        }
        this.StopProfiling("OvaleEnemies_RemoveEnemy");
    }
    private RemoveTaggedEnemy(cleuEvent, guid, timestamp) {
        this.StartProfiling("OvaleEnemies_RemoveTaggedEnemy");
        if (guid) {
            let name = self_enemyName[guid];
            let tagged = self_taggedEnemyLastSeen[guid];
            if (tagged) {
                self_taggedEnemyLastSeen[guid] = undefined;
                if (this.current.taggedEnemies > 0) {
                    this.current.taggedEnemies = this.current.taggedEnemies - 1;
                }
                this.DebugTimestamp("%s: %d/%d enemy removed: %s (%s), last tagged at %f", cleuEvent, this.current.taggedEnemies, this.current.activeEnemies, guid, name, tagged);
                Ovale.needRefresh();
            }
        }
        this.StopProfiling("OvaleEnemies_RemoveEnemy");
    }
    DebugEnemies() {
        for (const [guid, seen] of pairs(self_enemyLastSeen)) {
            let name = self_enemyName[guid];
            let tagged = self_taggedEnemyLastSeen[guid];
            if (tagged) {
                this.Print("Tagged enemy %s (%s) last seen at %f", guid, name, tagged);
            } else {
                this.Print("Enemy %s (%s) last seen at %f", guid, name, seen);
            }
        }
        this.Print("Total enemies: %d", this.current.activeEnemies);
        this.Print("Total tagged enemies: %d", this.current.taggedEnemies);
    }
    
    InitializeState() {
        this.next.enemies = undefined;
    }
    ResetState() {
        this.StartProfiling("OvaleEnemies_ResetState");
        this.next.activeEnemies = this.current.activeEnemies;
        this.next.taggedEnemies = this.current.taggedEnemies;
        this.StopProfiling("OvaleEnemies_ResetState");
    }
    CleanState() {
        this.next.activeEnemies = undefined;
        this.next.taggedEnemies = undefined;
        this.next.enemies = undefined;
    }
}

OvaleEnemies = new OvaleEnemiesClass();
OvaleState.RegisterState(OvaleEnemies);
